src/thor/triplegbuilder.cc

```diff
@@ -1208,8 +1208,14 @@ TripLegBuilder::Build(const AttributesController& controller,
     SetHeadings(trip_edge, controller, edge, shape, 0);
 
     auto* node = trip_path.add_node();
+    // when both start and end point are on the same edge, the duration of the leg should be
+    // set according to the difference of the percentage along the edge
+    //                start_node     end_node
+    // edge start =======0=============0==== edge end
+    //                   |_____________|
+    //                      duration
     if (controller.attributes.at(kNodeElapsedTime)) {
-      node->set_elapsed_time(path_begin->elapsed_time);
+      node->set_elapsed_time(path_begin->elapsed_time * (end_pct - start_pct));
```

Comment:
you cannot do this in all cases. remember that this code is used to build legs for both regular routing paths and for map matched routing paths.

for regular routing paths, the `PathInfo` objects `elapsed_time` is already trimmed to whatever the partial distance along the edge was. for map matching this is not the case.

i tried to put some pseudo code breadcrumbs here: https://github.com/valhalla/valhalla/compare/mm_timing